Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79170
* __->__ #78972
* #78523
* #78522
* #78677

Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)

eellison committed:
Make FakeTensors return meta within kerenl invocation, add FakeTensor op tests

[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta within kerenl invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors,  a couple meta registrations)





[ghstack-poisoned]

ezyang commented:
This doesn't invalidate the pull request but I want to make a higher level point: we are going to need support for dynamic shape soon, and this means that we were actually be using the C++ meta function implementations, for example we are not going to make tensor iterator symbolic aware. Instead we will be using python side implementations to get dynamic shape aware meta-tensor support. As a result it's not that important to make sure fake tensor works with tensor iterator (although it certainly is nice not to Segfault)

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -148,6 +149,8 @@ void autogradNotImplementedFallbackImpl(const c10::OperatorHandle& op, c10::Disp
   }, &stack_args_copy, 0, num_arguments);
   _foreach_tensor([&](size_t idx_tensor, size_t idx_ret, const at::Tensor& t) {
+    if (at::impl::tensor_has_dispatch(t) || at::impl::dispatch_mode_enabled())
+      return;
```
Tell me more about this one

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -14,6 +14,11 @@
 
+
+class ComplexInputException(Exception):
+    pass
```
Did you black this file, would be nice to have the formatting on its own

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -322,12 +391,15 @@ def check_non_fake_tensor(x):
                 if run_impl_check(func):
                     return op_impl(self, func, *args, **kwargs)
 
+            self.in_kernel_invocation = True
```
Naïvely I would've expected these to pair up with the no dispatch call

eellison replied:
do you mean put inside the `no_dispatch` or to use `_DisableTorchDispatch` instead of `self.in_kernel_invocation` ? I haven't thought too hard about it... maybe would be a good cleanup, although I still need to do a little more thinking/understanding of composability

eellison commented:
@ezyang good point. there were a lot of issues not just within TE, though, such as `copy_` `clone`, the `linalg` test failures, Distribution error checking, etc, etc. 

Other seemingly innocuous code like `aten::outer` could also lead to issues: 

[self.reshape({self.size(0), 1}) * vec2](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebra.cpp#L1155) within no_dispatch would break because `self.reshape` would get confused at `self` having a different device (cpu) than dispatch key...

Relatedly, it would be great to not have to handle something as simple as `aten::outer` in the a TorchDispatchMode but right now there is no way to decompose it (i will write up more full issue). If a decomposition for it exists in python, then you can just run that, but there is no equivalent way to take a composite kernel in C++ and run each operator with the mode enabled. If you don't wrap the kernel with `no_dispatch` it will just endlessly recur. I had some luck running the decompositions defined in python for this but coverage isn't all of the way there and it proved unnecessary at least for now to get the tests passing.

ezyang commented:
I'm not sure what the best way to get the test passing is, but our plan on record is to get all of the composites rewritten in python

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison committed:
Update on "Make FakeTensors return meta device within kernel invocation, add FakeTensor op tests"


Add FakeTensor Op tests, as well as necessary changes to make the tests pass. 

- Previously, `FakeTensor(cpu, ...)` would always return `cpu` if `device` was called. While this the behavior you want within userland code, within the kernel itself you want to compute as if everything is a meta tensor. For instance, within TensorIterator there are a number of `is_meta()` checks that before this PR would return False, and attempt to proceed as if the FakeTensors were on CPU instead of meta, which led to segfaults among other incorrect behavior. [Another example is in _linalg_check_errors](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1298). This PR changes `FakeTensorMode` to store whether or not we are in a kernel invocation, and while that is True, return `meta`. 

- Extends the error checking in `VariableType_*` to ignore Tensors with torch_dispatch defined, in addition to the existing when a torch_mode is defined. 

- A few small misc changes (adding a couple new ops to be special-handled in FakeTensors)





[ghstack-poisoned]

eellison commented:
@pytorchbot merge
